Modified channel creating logic to accept no alphanumerical only chan…

gomessguii opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
Changed the logic for creating a new channel to generate a random 32 characters long string for the *Channel.name* (same as url) whenever the display name is filled but there's no url-safe (alphanumeric) characters in it. 

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-50305

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
None

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

|  before  |  after  |
|----|----|
| ![índice](https://user-images.githubusercontent.com/80290114/217154338-73959d3e-1fdb-42df-9601-bdef54dcc4b1.png) | ![Captura de tela de 2023-02-07 01-31-24](https://user-images.githubusercontent.com/80290114/217154412-2db65887-c1c8-4d15-af0e-d5e74cab96c3.jpg) |



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Implemented url auto generation on channel creation for when there's no url safe characters on its name
```

mattermost-build commented:
Hello @gomessguii,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

hmhealey commented on webapp/package.json:
```diff
@@ -13,6 +13,7 @@
     "clean": "npm run clean --workspaces --if-present"
   },
   "devDependencies": {
+    "@types/crypto-js": "4.1.1",
```
@gomessguii This should actually be added to `webapp/channels/package.json`. This one at the top level is just for the scripts to run all the different packages in the monorepo

gomessguii replied:
I just ran npm install, I wasn't aware of the directories scheme. I'll fix it.

hmhealey replied:
Ah, I see. I should've recommended running `npm add --workspace=channels`. I've submitted a PR to explain that properly since I realized I never explained how to do that in any of our docs

hmhealey commented on webapp/package.json:
```diff
@@ -25,5 +26,5 @@
     "playbooks",
     "boards"
-  ]
+  ],
```
This extra comma is actually causing the `npm install` to fail for this. Some JSON code allows for trailing commas, but strictly speaking, they're not supported in JSON

gomessguii replied:
It didn't fail on my PC and I guess the comma was added by the npm command or by VSCode maybe. 
I'll fix it anyway.

hmhealey commented on webapp/channels/src/components/new_channel_modal/new_channel_modal.tsx:
```diff
@@ -8,6 +8,8 @@ import {Tooltip} from 'react-bootstrap';
 
 import classNames from 'classnames';
 
+import { random as cryptoJSRandomWordArray} from 'crypto-js/lib-typedarrays';
```
```suggestion
import {random as cryptoJSRandomWordArray} from 'crypto-js/lib-typedarrays';
```

gomessguii replied:
fixed

hmhealey reviewed (changes requested):
Looks good to me. There's just one thing that's failing the style check on CI

hmhealey commented:
Looks like you forgot to re-add `crypto-js` to the dependencies. Mind adding that?

gomessguii commented:
> Looks like you forgot to re-add `crypto-js` to the dependencies. Mind adding that?

I don't understand, I removed it from  `webapp/package.json` as requested and added it on line 125 of `webapp/channels/package.json`. I ran the app locally and worked just fine. 
Was the lib supposed to be in both files?

hmhealey commented:
You re-added `@types/crypto-js` to the `devDependencies` in `webapp/channels/package.json`, but you forgot to also re-add `crypto-js` to the `dependencies` section there, so we've got the type definitions for the library but not the code. Running locally, you probably have the previously installed version in the `node_modules` since that's not removed automatically

gomessguii commented:
> You re-added `@types/crypto-js` to the `devDependencies` in `webapp/channels/package.json`, but you forgot to also re-add `crypto-js` to the `dependencies` section there, so we've got the type definitions for the library but not the code. Running locally, you probably have the previously installed version in the `node_modules` since that's not removed automatically

You're right. It was in the node_modules. I should've checked it. Sorry about that.
I fixed the dependency, and this time, I deleted the modules folder and ran `npm install` again to make sure.

M-ZubairAhmed commented on webapp/channels/package.json:
```diff
@@ -34,6 +34,7 @@
     "core-js": "3.26.0",
     "country-list": "2.2.0",
     "crypto-browserify": "3.12.0",
+    "crypto-js": "4.1.1",
```
We already have a `crypto-browserify` library. Can you try to use that instead? Agreed, it might not have straight forward APIs for generating words array. Let me know if you need help with that. :)

gomessguii replied:
It's possible. However it will require some changes in the implementation.
At first, I had trouble importing the lib with usual react import. Typescript required `@types/crypto-browserify` and I keep getting 404 when I run `npm install --save-dev @types/crypto-browserify` . 
To bypass that, I imported it using nodejs require. Like this:
~~~ ts
const crypto = require('crypto');
~~~
Please let me know if you guys don't see a problem on importing like that or if I missed something in the regular import that got me those problems.

The behavior is exactly the same and the implementation changes to this:
~~~ ts
setURL(crypto.randomBytes(16).toString('hex'));
~~~

M-ZubairAhmed replied:
You are absolutely correct we have aliased in `webpack.config.js`. There are other places in codebase which imports `crypto` as well. The above implementation looks good to me i checked out locally. Would you like to proceed with that

gomessguii replied:
Yes. I just pushed the commit with that change.

asaadmahmood reviewed (dismissed):
LGTM from a UX perspective!

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented on webapp/channels/src/components/new_channel_modal/new_channel_modal.tsx:
```diff
@@ -104,6 +104,8 @@ const NewChannelModal = () => {
     const [canCreateFromPluggable, setCanCreateFromPluggable] = useState(true);
     const [actionFromPluggable, setActionFromPluggable] = useState<((currentTeamId: string, channelId: string) => Promise<Board>) | undefined>(undefined);
 
+    const crypto = require('crypto');
```
you can import just like https://github.com/mattermost/mattermost-server/blob/8f9b22b004ca53d0becbce64edf171ae0eda17da/webapp/channels/src/components/admin_console/generated_setting.tsx#L3-L4

M-ZubairAhmed reviewed (approved):
It looks perfect thanks for making the changes.

hmhealey commented:
I'm going to update the branch for this PR again to see if that fixes the EE tests since this obviously wouldn't have had an effect on them

/update-branch

stevemudie commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/330820) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/330820)

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

M-ZubairAhmed commented:
/update-branch

yasserfaraazkhan commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/336166) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/336166)

yasserfaraazkhan commented:
/e2e-test

yasserfaraazkhan commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/336303) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/336303)

cwarnermm commented:
@hmhealey - Is there any outstanding documentation needed for this PR? Or can we safely change the label from **Docs/Needed** to **Docs/Done**?

hmhealey commented:
I don't think we need to. This just changes it so that the user isn't forced to manually make the channel URL if the channel's name is entirely non-Latin characters
